modules/grpc/otel/otel-dest-worker.cpp

```diff
@@ -56,16 +76,325 @@ DestWorker::disconnect()
+    {
+      ResourceSpans *possible_resource_spans = trace_service_request.mutable_resource_spans(i);
+      if (MessageDifferencer::Equals(possible_resource_spans->resource(), current_msg_metadata.resource) &&
+          possible_resource_spans->schema_url() == current_msg_metadata.resource_schema_url)
+        {
+          resource_spans = possible_resource_spans;
+          break;
+        }
+    }
+  if (!resource_spans)
+    {
+      resource_spans = trace_service_request.add_resource_spans();
+      resource_spans->mutable_resource()->CopyFrom(current_msg_metadata.resource);
+      resource_spans->set_schema_url(current_msg_metadata.resource_schema_url);
+    }
+
+  ScopeSpans *scope_spans = nullptr;
+  for (int i = 0; i < resource_spans->scope_spans_size(); i++)
+    {
+      ScopeSpans *possible_scope_spans = resource_spans->mutable_scope_spans(i);
+      if (MessageDifferencer::Equals(possible_scope_spans->scope(), current_msg_metadata.scope) &&
+          possible_scope_spans->schema_url() == current_msg_metadata.scope_schema_url)
+        {
+          scope_spans = possible_scope_spans;
+          break;
+        }
+    }
+  if (!scope_spans)
+    {
+      scope_spans = resource_spans->add_scope_spans();
+      scope_spans->mutable_scope()->CopyFrom(current_msg_metadata.scope);
+      scope_spans->set_schema_url(current_msg_metadata.scope_schema_url);
+    }
+
+  return scope_spans;
+}
+
+bool
+DestWorker::insert_log_record_from_log_msg(LogMessage *msg)
+{
+  ScopeLogs *scope_logs = lookup_scope_logs(msg);
+  if (!scope_logs)
+    return false;
+
+  LogRecord *log_record = scope_logs->add_log_records();
+  return formatter.format(msg, *log_record);
+}
+
+bool
+DestWorker::insert_fallback_log_record_from_log_msg(LogMessage *msg)
+{
+  ScopeLogs *scope_logs = lookup_scope_logs(msg);
+  if (!scope_logs)
+    return false;
+
+  LogRecord *log_record = scope_logs->add_log_records();
+  return formatter.format_fallback(msg, *log_record);
+}
+
+bool
+DestWorker::insert_metric_from_log_msg(LogMessage *msg)
+{
+  ScopeMetrics *scope_metrics = lookup_scope_metrics(msg);
+  if (!scope_metrics)
+    return false;
+
+  Metric *metric = scope_metrics->add_metrics();
+  return formatter.format(msg, *metric);
+}
+
+bool
+DestWorker::insert_span_from_log_msg(LogMessage *msg)
+{
+  ScopeSpans *scope_spans = lookup_scope_spans(msg);
+  if (!scope_spans)
+    return false;
+
+  Span *span = scope_spans->add_spans();
+  return formatter.format(msg, *span);
+}
+
 LogThreadedResult
 DestWorker::insert(LogMessage *msg)
 {
-  return LTR_SUCCESS;
+  MessageType type = get_message_type(msg);
+  switch (type)
+    {
+    case MessageType::LOG:
+      if (!insert_log_record_from_log_msg(msg))
+        return LTR_ERROR;
+      break;
+    case MessageType::METRIC:
+      if (!insert_metric_from_log_msg(msg))
+        return LTR_ERROR;
+      break;
+    case MessageType::SPAN:
+      if (!insert_span_from_log_msg(msg))
+        return LTR_ERROR;
```

Comment:
Instead of producing `LTR_ERROR`, we should use the fallback log record when we have non-transient errors.

An alternative would be `LTR_DROP`, but I believe we should have a fallback method for both non-otel-originated messages and problematic otel-originated messages.

In the second case, if we can't provide a good fallback (or we can't send a partially formatted message), we can maybe consider using `LTR_DROP`.

Reply:
As we talked about this IRL, we cannot really define a failed format of a Span to fit in a LogRecord, so modified the code to DROP in that case.